plugins/glustershd/rest.go

```diff
@@ -0,0 +1,198 @@
+package glustershd
+
+import (
+	"net/http"
+
+	"github.com/gluster/glusterd2/glusterd2/gdctx"
+	restutils "github.com/gluster/glusterd2/glusterd2/servers/rest/utils"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/pkg/api"
+	"github.com/gluster/glusterd2/pkg/errors"
+
+	"github.com/gorilla/mux"
+	"github.com/pborman/uuid"
+	log "github.com/sirupsen/logrus"
+)
+
+func isVolReplicate(vType volume.VolType) bool {
+	if vType == volume.Replicate || vType == volume.Disperse || vType == volume.DistReplicate || vType == volume.DistDisperse {
+		return true
+	}
+
+	return false
+}
+
+func glustershEnableHandler(w http.ResponseWriter, r *http.Request) {
+	// Implement the help logic and send response back as below
+	p := mux.Vars(r)
+	volname := p["name"]
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	//validate volume name
+	v, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound.Error(), api.ErrCodeDefault)
+		return
+	}
+	// Store initial volinfo before changing the HealFlag
+	oldvolinfo := v
+
+	// validate volume type
+	if !isVolReplicate(v.Type) {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Volume Type not supported", api.ErrCodeDefault)
+		return
+	}
+
+	if v.State != volume.VolStarted {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Volume should be in started state.", api.ErrCodeDefault)
+		return
+
+	}
+
+	// Transaction which starts self heal daemon on all nodes with atleast one brick.
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+
+	//Lock on Volume Name
+	lock, unlock, err := transaction.CreateLockSteps(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err.Error(), api.ErrCodeDefault)
+		return
+	}
+
+	v.HealEnabled = true
+
+	txn.Nodes = v.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc:   "vol-option.UpdateVolinfo",
+			Nodes:    []uuid.UUID{gdctx.MyUUID},
+			UndoFunc: "selfheal-start-undo",
```

Comment:
The UndoFunc shouldn't be the same for both steps. It is good enough if you specify it for this step.